test/healthcheck_test.go

```diff
@@ -653,6 +686,27 @@ func TestHealthCheckWithoutReportHealthCalledAddrConnShutDown(t *testing.T) {
 // onGoAway/onClose goroutine.
 func TestHealthCheckWithoutReportHealthCalled(t *testing.T) {
 	defer leakcheck.Check(t)
+
+	hcEnterChan := make(chan struct{})
+	hcExitChan := make(chan struct{})
+	testHealthCheckFuncWrapper := func(ctx context.Context, newStream func() (interface{}, error), update func(bool), service string) error {
+		select {
+		case <-hcEnterChan:
+			return nil
+		default:
+			close(hcEnterChan)
+		}
+		err := testHealthCheckFunc(ctx, newStream, update, service)
+		close(hcExitChan)
+		return err
+	}
+
+	replace := replaceHealthCheckFunc(testHealthCheckFuncWrapper)
+	defer replace()
+	// this leakcheck is used here to make sure grpc goroutines have all exited so we can safely replace
+	// the health check func without data race.
+	defer leakcheck.Check(t)
```

Comment:
Double-leakcheck.

These tests all look very similar in setup at least.  Can we consolidate somehow?  (Table-driven or common setup/check functions.)